File: frontend/src/types/autocomplete.ts
/**
 * Autocomplete Types
 *
 * Type definitions for the autocomplete API and components.
 */

/**
 * Metadata for a user autocomplete suggestion
 */
export interface AutocompleteSuggestionMeta {
  /** User ID (for recent searches) */
  userId?: number;
  /** User's profile picture URL */
  profilePic?: string;
  /** Whether the user is verified */
  isVerified?: boolean;
  /** User's follower count */
  followersCount?: number;
}

/**
 * A single autocomplete suggestion
 */
export interface AutocompleteSuggestion {
  /** Display text (username) */
  text: string;
  /** Type of suggestion */
  kind: 'user' | 'query' | 'video' | 'entity';
  /** Relevance score (higher = more relevant) */
  score: number;
  /** Additional metadata */
  meta?: AutocompleteSuggestionMeta;
}

/**
 * API response for autocomplete endpoint
 */
export interface AutocompleteResponse {
  /** Echo of the original query */
  query: string;
  /** Unique request ID for tracing */
  requestId: string;
  /** Ranked list of suggestions */
  suggestions: AutocompleteSuggestion[];
}

/**
 * A user in search suggestions (recent or trending)
 */
export interface SearchSuggestionUser {
  /** User ID */
  id: number;
  /** Username */
  username: string;
  /** Profile picture URL */
  profilePic?: string;
  /** Whether the user is verified */
  isVerified: boolean;
  /** Follower count */
  followersCount: number;
}

/**
 * API response for search suggestions endpoint
 * Returns recent searches and trending users
 */
export interface SearchSuggestionsResponse {
  /** User's recent profile searches */
  recent: SearchSuggestionUser[];
  /** Personalized trending users */
  trending: SearchSuggestionUser[];
}

File: backend/internal/dto/autocomplete.go
package dto

// ==================== Autocomplete DTOs ====================

// AutocompleteSuggestionMeta contains additional metadata for a suggestion
// @Description Additional metadata for autocomplete suggestion (profile pic, verified status, etc.)
type AutocompleteSuggestionMeta struct {
	ProfilePic     *string `json:"profilePic,omitempty" example:"https://storage.blob.core.windows.net/pics/1/abc.jpg"`
	IsVerified     bool    `json:"isVerified" example:"false"`
	FollowersCount int64   `json:"followersCount" example:"150"`
}

// AutocompleteSuggestion represents a single autocomplete suggestion
// @Description A single autocomplete suggestion with text, kind, score, and metadata
type AutocompleteSuggestion struct {
	Text  string                     `json:"text" example:"john_doe"`
	Kind  string                     `json:"kind" example:"user"`
	Score float64                    `json:"score" example:"85.5"`
	Meta  AutocompleteSuggestionMeta `json:"meta"`
}

// AutocompleteResponse represents the autocomplete API response
// @Description Autocomplete response with query echo, request ID for tracing, and ranked suggestions
type AutocompleteResponse struct {
	Query       string                   `json:"query" example:"john"`
	RequestID   string                   `json:"requestId" example:"550e8400-e29b-41d4-a716-446655440000"`
	Suggestions []AutocompleteSuggestion `json:"suggestions"`
}
